### PR TITLE
feat: add Stargate logics

### PR DESCRIPTION
**Pull request by Jeff-CCH**
- [x] add Stargate
- [x] uni test
- [x] e2e test
- [x] yarn changeset

### CODE REVIEW

File: src/logics/index.ts [hunk 1]
```diff
@@ -13,3 +13,4 @@ export * as syncswap from './syncswap';
 export * as uniswapv3 from './uniswap-v3';
 export * as utility from './utility';
 export * as zeroexv4 from './zeroex-v4';
+export * as stargate from './stargate';
```
Comment (zodahu) [inline src/logics/index.ts:16]: It was arranged in alphabetical order.

File: src/logics/stargate/configs.ts [hunk 2]
```diff
@@ -0,0 +1,537 @@
...
+
+export interface Config {
+  id: number;
+  chainId: number;
+  contract: Record<ContractNames, string>;
+  pools: Pool[];
+  STG?: common.Token;
+}
+
+export const configs: Config[] = [
```
Comment (zodahu) [inline src/logics/stargate/configs.ts:50]: The configs are lack of `ChainId.BASE`

File: src/logics/stargate/configs.ts [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+          { chainId: common.ChainId.avalanche, poolIds: [PoolId.USDC, PoolId.USDT] },
+        ],
+      },
+      {
+        id: PoolId.USDT,
+        token: mainnetTokens.USDT,
+        decimals: 6,
+        paths: [
+          { chainId: common.ChainId.optimism, poolIds: [PoolId.USDC] },
+          { chainId: common.ChainId.polygon, poolIds: [PoolId.USDC, PoolId.USDT] },
```
Comment (zodahu) [inline src/logics/stargate/configs.ts:79]: Curious that if it's same between `[PoolId.USDC, PoolId.USDT]` and `[PoolId.USDT, PoolId.USDC]`, as furucombo-interface is `poolIds: [2, 1]`

File: src/logics/stargate/logic.swap-token.ts [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+      const amountOffset = balanceBps ? common.getParamOffset(2) : undefined;
+
+      inputs = [
+        core.newLogicInput({
+          input: new common.TokenAmount(input.token.wrapped, input.amount),
+          balanceBps,
+          amountOffset,
+        }),
+        core.newLogicInput({ input: new common.TokenAmount(getNativeToken(this.chainId)).setWei(fee.toString()) }),
```
Comment (zodahu) [inline src/logics/stargate/logic.swap-token.ts:139]:
```suggestion
        core.newLogicInput({ input: new common.TokenAmount(getNativeToken(this.chainId)).setWei(fee) }),
```

File: src/logics/stargate/logic.swap-token.ts [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+      const feeLibrary = FeeLibrary__factory.connect(getContractAddress(this.chainId, 'FeeLibrary'), this.provider);
+      const { eqFee, eqReward, lpFee, protocolFee } = await feeLibrary.getFees(
+        srcPoolId,
+        dstPoolId,
+        getStargateChainId(dstChainId),
+        receiver,
+        amountIn
+      );
+      const totalFee = eqFee.add(lpFee).add(protocolFee);
+      output = new common.TokenAmount(tokenOut).setWei(amountIn.sub(totalFee).add(eqReward));
```
[the remark this answers is not in the document]
Comment (Jeff-CCH) [reply src/logics/stargate/logic.swap-token.ts:95]: IMO, this is for adjusting the `amountOut` for different decimals. But the decimal is handled in tokenAmount by `setWei()`.

File: src/logics/stargate/configs.ts (hunk 3 above)
Comment (Jeff-CCH) [reply src/logics/stargate/configs.ts:79]: IMO, they are the same. But we can double-check this one with the author.

File: src/logics/stargate/logic.swap-token.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Jeff-CCH) [reply src/logics/stargate/logic.swap-token.ts:95]: Please Jay help revisit this to see if there is anything we need to change.

File: src/logics/stargate/configs.ts (hunk 3 above)
Comment (dinngodev) [reply src/logics/stargate/configs.ts:88]: The order will only affect the display of the Furucombo interface. In our interface, we assume that the user only wants to bridge, not bridge and swap. So we can find out that the first pool in the poolIds always matches the token.

Comment (weizard) [reply src/logics/stargate/configs.ts:88]: The order will only affect the display of the Furucombo interface. In our interface, we assume that the user only wants to bridge, not bridge and swap. So we can find out that the first pool in the poolIds always matches the token.

Comment (zodahu) [reply src/logics/stargate/configs.ts:88]: @Jeff-CCH I think we can follow the same approach as Furucombo as the UI may use the same logic to display the token list after migration.

File: src/logics/stargate/configs.ts [hunk 6]
```diff
@@ -0,0 +1,592 @@
...
+    pools: [
+      {
+        id: PoolId.USDC,
+        token: baseTokens.USDbC,
+        decimals: 6,
+        paths: [
+          { chainId: common.ChainId.mainnet, poolIds: [PoolId.USDC, PoolId.USDT] },
+          { chainId: common.ChainId.optimism, poolIds: [PoolId.USDC] },
+          { chainId: common.ChainId.polygon, poolIds: [PoolId.USDC] },
+          { chainId: common.ChainId.arbitrum, poolIds: [PoolId.USDC] },
```
Comment (zodahu) [inline src/logics/stargate/configs.ts:346]:
```suggestion
          { chainId: common.ChainId.polygon, poolIds: [PoolId.USDC, PoolId.USDT] },
          { chainId: common.ChainId.arbitrum, poolIds: [PoolId.USDC, PoolId.USDT] },
```

Comment (Jeff-CCH) [reply src/logics/stargate/configs.ts:346]: from the Stargate UI, I didn't see the mapping. Can you illustrate where do you find it?

Comment (zodahu) [reply src/logics/stargate/configs.ts:346]: Let's follow the UI as it seems their [docs](https://stargateprotocol.gitbook.io/stargate/developers/stargate-chain-paths) isn't up to date, thanks.

**Pull request merged**